crates/apub/src/fetcher/search.rs

```diff
@@ -46,88 +44,4 @@ pub(crate) async fn search_query_to_object_id_local(
-  PersonOrCommunity(Box<UserOrCommunity>),
-}
-
-#[derive(Deserialize)]
-#[serde(untagged)]
-pub(crate) enum SearchableKinds {
-  PageOrNote(Box<PageOrNote>),
-  PersonOrGroup(Box<PersonOrGroup>),
-}
-
-#[async_trait::async_trait]
-impl Object for SearchableObjects {
-  type DataType = LemmyContext;
-  type Kind = SearchableKinds;
-  type Error = LemmyError;
-
-  fn last_refreshed_at(&self) -> Option<DateTime<Utc>> {
-    match self {
-      SearchableObjects::PostOrComment(p) => p.last_refreshed_at(),
-      SearchableObjects::PersonOrCommunity(p) => p.last_refreshed_at(),
-    }
-  }
-
-  // TODO: this is inefficient, because if the object is not in local db, it will run 4 db queries
-  //       before finally returning an error. it would be nice if we could check all 4 tables in
-  //       a single query.
-  //       we could skip this and always return an error, but then it would always fetch objects
-  //       over http, and not be able to mark objects as deleted that were deleted by remote server.
-  async fn read_from_id(
-    object_id: Url,
-    context: &Data<Self::DataType>,
-  ) -> LemmyResult<Option<Self>> {
-    let uc = UserOrCommunity::read_from_id(object_id.clone(), context).await?;
-    if let Some(uc) = uc {
-      return Ok(Some(SearchableObjects::PersonOrCommunity(Box::new(uc))));
-    }
-    let pc = PostOrComment::read_from_id(object_id.clone(), context).await?;
-    if let Some(pc) = pc {
-      return Ok(Some(SearchableObjects::PostOrComment(Box::new(pc))));
-    }
-    Ok(None)
-  }
-
-  async fn delete(self, data: &Data<Self::DataType>) -> LemmyResult<()> {
-    match self {
-      SearchableObjects::PostOrComment(pc) => pc.delete(data).await,
-      SearchableObjects::PersonOrCommunity(pc) => pc.delete(data).await,
-    }
-  }
-
-  async fn into_json(self, data: &Data<Self::DataType>) -> LemmyResult<Self::Kind> {
-    use SearchableObjects::*;
-    Ok(match self {
-      PostOrComment(pc) => SearchableKinds::PageOrNote(Box::new(pc.into_json(data).await?)),
-      PersonOrCommunity(pc) => SearchableKinds::PersonOrGroup(Box::new(pc.into_json(data).await?)),
-    })
-  }
-
-  async fn verify(
-    apub: &Self::Kind,
-    expected_domain: &Url,
-    data: &Data<Self::DataType>,
-  ) -> LemmyResult<()> {
-    use SearchableKinds::*;
-    match apub {
-      PageOrNote(pn) => PostOrComment::verify(pn, expected_domain, data).await,
-      PersonOrGroup(pg) => UserOrCommunity::verify(pg, expected_domain, data).await,
-    }
-  }
-
-  async fn from_json(apub: Self::Kind, context: &Data<LemmyContext>) -> LemmyResult<Self> {
-    use SearchableKinds::*;
-    use SearchableObjects as SO;
-    Ok(match apub {
-      PageOrNote(pg) => SO::PostOrComment(Box::new(PostOrComment::from_json(*pg, context).await?)),
-      PersonOrGroup(pg) => {
-        SO::PersonOrCommunity(Box::new(UserOrCommunity::from_json(*pg, context).await?))
-      }
-    })
-  }
-}
+pub(crate) type SearchableObjects = Either<PostOrComment, UserOrCommunity>;
```

Comment:
It'd be nice if this could be flattened somehow.

Reply:
There are different `Either` types which provide `Either3`, `Either4` etc, but unfortunately the either crate doesnt have that.